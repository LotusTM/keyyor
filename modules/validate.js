const t = require('tcomb-validation')

/**
 * Validate value against type with tcomb-validation and print
 * nice list of errors in case of fail
 * @param  {*}        value Value to be validated
 * @param  {function} type  tcomb Type to validate value against
 * @return {object} Validated tcomb object with all `tcomb-validation` methods
 * @throws {TypeError} If there are any validation errors or if type isn't tcomb type
 * @example
 *   validate({ id: 'testid', number: 123 }, t.struct({ id: t.String, number: t.Number }))).isValid() // => true
 */
module.exports = (value, type) => {
  const validated = t.Type(type) && t.validate(value, type, { path: [type.displayName] })
  const errors = validated.errors

  if (errors.length) {
    const message = errors.map((e) => e.message).reduce((acc, cur) => acc + '\n' + cur)
    throw new TypeError(`[tcomb]\n${message}`)
  }

  return validated
}
