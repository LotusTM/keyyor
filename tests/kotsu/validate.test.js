/* eslint-env jest */

import t from 'tcomb'
import validate from '../../modules/validate'

describe('Tcomb validate', () => {
  const ValidateTestSchema = t.struct({
    id: t.String,
    number: t.Number
  })

  it('should pass valid data', () => {
    expect(validate({ id: 'testid', number: 123 }, ValidateTestSchema)).toMatchSnapshot()
  })
  it('should error on ivalid data', () => {
    expect(() => validate({ nope: 'wrongprop' }, ValidateTestSchema)).toThrowErrorMatchingSnapshot()
    expect(() => validate({ nope: 'wrongprop' }, 'Not a tcomb type')).toThrowErrorMatchingSnapshot()
  })
})
