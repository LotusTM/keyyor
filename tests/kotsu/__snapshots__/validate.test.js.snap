// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tcomb validate should error on ivalid data 1`] = `
"[tcomb]
Invalid value undefined supplied to /Struct{id: String, number: Number}/id: String
Invalid value undefined supplied to /Struct{id: String, number: Number}/number: Number"
`;

exports[`Tcomb validate should error on ivalid data 2`] = `"[tcomb] Invalid value \\"Not a tcomb type\\" supplied to Type"`;

exports[`Tcomb validate should pass valid data 1`] = `
Struct {
  "errors": Array [],
  "value": Struct {
    "id": "testid",
    "number": 123,
  },
}
`;
